iwyu_output.cc

```diff
@@ -1889,6 +1893,41 @@ size_t PrintableDiffs(const string& filename,
 
 }  // namespace internal
 
+void IwyuFileInfo::HandlePreprocessingDone() {
+  // Check macros defined by includer.  Requires file preprocessing to be
+  // finished to know all direct includes and all macro usages.
+  bool should_report_violations = ShouldReportIWYUViolationsFor(file_);
+  std::list<const FileEntry*> direct_macro_use_includees;
+  std::set_intersection(macro_users_.begin(), macro_users_.end(),
+                        direct_includes_as_fileentries_.begin(),
+                        direct_includes_as_fileentries_.end(),
+                        std::inserter(direct_macro_use_includees,
+                                      direct_macro_use_includees.end()));
+  for (const FileEntry* macro_use_includee : direct_macro_use_includees) {
+    if (should_report_violations) {
+      ERRSYM(file_) << "Keep #include " << macro_use_includee->getName()
```

Comment:
OK, so the mere presence of a dependency on a includer-defined macro now forces us to keep the include?

I think that may be too strong, e.g. for 'control' macros, something like:

     #define XYZ_ENABLE_FOOBAR
     #include "unused.h"

where `unused.h` has something like:

    #ifdef XYZ_ENABLE_FOOBAR
    inline void foobar() {
    }
    #endif

(practical example: `__STDC_LIMIT_MACROS` and `__STDC_CONSTANT_MACROS` vs `stdint.h`).

That said, in practice I think this is fine. It will produce slightly less optimal IWYU output (for a narrow case), but should always produce code that can compile.